staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go

```diff
@@ -273,122 +368,98 @@ func (cnc *CloudNodeController) updateNodeAddress(ctx context.Context, node *v1.
-				Type:               v1.NodeNetworkUnavailable,
-				Status:             v1.ConditionTrue,
-				Reason:             "NoRouteCreated",
-				Message:            "Node created without a route",
-				LastTransitionTime: metav1.Now(),
-			}); err != nil {
-				return err
-			}
-		}
-		return nil
-	})
+// syncNode handles updating existing nodes registered with the cloud taint
+// and processes nodes that were added into the cluster, and cloud initialize them if appropriate.
+func (cnc *CloudNodeController) syncNode(ctx context.Context, nodeName string) error {
+	curNode, err := cnc.nodeInformer.Lister().Get(nodeName)
 	if err != nil {
-		utilruntime.HandleError(err)
-		return
-	}
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
 
-	curNode, err := cnc.kubeClient.CoreV1().Nodes().Get(context.TODO(), node.Name, metav1.GetOptions{})
-	if err != nil {
-		utilruntime.HandleError(fmt.Errorf("failed to get node %s: %v", node.Name, err))
-		return
+		return err
 	}
 
 	cloudTaint := getCloudTaint(curNode.Spec.Taints)
 	if cloudTaint == nil {
 		// Node object received from event had the cloud taint but was outdated,
-		// the node has actually already been initialized.
-		return
+		// the node has actually already been initialized, so this sync event can be ignored.
+		return nil
 	}
 
-	providerID, err := cnc.getProviderID(ctx, curNode)
+	klog.Infof("Initializing node %s with cloud provider", nodeName)
+
+	copyNode := curNode.DeepCopy()
+	providerID, err := cnc.getProviderID(ctx, copyNode)
 	if err != nil {
-		utilruntime.HandleError(fmt.Errorf("failed to get provider ID for node %s at cloudprovider: %v", node.Name, err))
-		return
+		return fmt.Errorf("failed to get provider ID for node %s at cloudprovider: %v", nodeName, err)
 	}
 
-	instanceMetadata, err := cnc.getInstanceMetadata(ctx, providerID, curNode)
+	instanceMetadata, err := cnc.getInstanceMetadata(ctx, providerID, copyNode)
 	if err != nil {
-		utilruntime.HandleError(fmt.Errorf("failed to get instance metadata for node %s: %v", node.Name, err))
-		return
+		return fmt.Errorf("failed to get instance metadata for node %s: %v", nodeName, err)
 	}
 
-	nodeModifiers, err := cnc.getNodeModifiersFromCloudProvider(ctx, providerID, curNode, instanceMetadata)
+	nodeModifiers, err := cnc.getNodeModifiersFromCloudProvider(ctx, providerID, copyNode, instanceMetadata)
 	if err != nil {
-		utilruntime.HandleError(fmt.Errorf("failed to initialize node %s at cloudprovider: %v", node.Name, err))
-		return
+		return fmt.Errorf("failed to get node modifiers from cloud provider: %v", err)
 	}
 
 	nodeModifiers = append(nodeModifiers, func(n *v1.Node) {
 		n.Spec.Taints = excludeCloudTaint(n.Spec.Taints)
 	})
 
 	err = clientretry.RetryOnConflict(UpdateNodeSpecBackoff, func() error {
-		curNode, err := cnc.kubeClient.CoreV1().Nodes().Get(context.TODO(), node.Name, metav1.GetOptions{})
-		if err != nil {
-			return err
+		var curNode *v1.Node
+		if cnc.cloud.ProviderName() == "gce" {
```

Comment:
FYI @cheftako we should probably discuss getting rid of this custom check for GCE, I can't recall the initial reason behind the GCE specific check. Would probably be nice to have something generic here.

Reply:
+1. Most of these checks are because the test needs something which wasn't generally available. Not sure about this particular test but I've written a few of these. Its frequently something like the webhook tests. You need to create a special test container and deploy it to your registry. The intention was always that other cloud providers would follow adding the container to their registry.